lib/SimpleImplicitDiscreteSolve/src/SimpleImplicitDiscreteSolve.jl

```diff
@@ -0,0 +1,95 @@
+module SimpleImplicitDiscreteSolve
+
+using SciMLBase
+using SimpleNonlinearSolve
+using Reexport
+@reexport using DiffEqBase
+
+"""
+    SimpleIDSolve()
+
+Simple solver for `ImplicitDiscreteSystems`. Uses `SimpleNewtonRaphson` to solve for the next state at every timestep.
+"""
+struct SimpleIDSolve <: SciMLBase.AbstractODEAlgorithm end
+
+function DiffEqBase.__init(prob::ImplicitDiscreteProblem, alg::SimpleIDSolve; dt = 1)
+    u0 = prob.u0
+    p = prob.p
+    f = prob.f
+    t = prob.tspan[1]
+
+    nlf = isinplace(f) ? (out, u, p) -> f(out, u, u0, p, t) : (u, p) -> f(u, u0, p, t)
+    prob = NonlinearProblem{isinplace(f)}(nlf, u0, p)
+    sol = solve(prob, SimpleNewtonRaphson())
+    sol, (sol.retcode != ReturnCode.Success)
+end
+
+function DiffEqBase.solve(prob::ImplicitDiscreteProblem, alg::SimpleIDSolve; 
+    dt = 1,
+    save_everystep = true,
+    save_start = true,
+    adaptive = false,
+    dense = false,
+    save_end = true,
+    kwargs...)
+
+    @assert !adaptive
+    @assert !dense
+    (initsol, initfail) = DiffEqBase.__init(prob, alg; dt)
+    if initfail
+        sol = DiffEqBase.build_solution(prob, alg, prob.tspan[1], u0, k = nothing, stats = nothing, calculate_error = false) 
+        return SciMLBase.solution_new_retcode(sol, ReturnCode.InitialFailure)
+    end
+
+    u0 = initsol.u
+    tspan = prob.tspan
+    f = prob.f
+    p = prob.p
+    t = tspan[1]
+    tf = prob.tspan[2]
+    ts = tspan[1]:dt:tspan[2]
+
+    if save_everystep && save_start
+        us = Vector{typeof(u0)}(undef, length(ts))
+        us[1] = u0
+    elseif save_everystep
+        us = Vector{typeof(u0)}(undef, length(ts) - 1)
+    elseif save_start
+        us = Vector{typeof(u0)}(undef, 2)
+        us[1] = u0
+    else
+        us = Vector{typeof(u0)}(undef, 1) # for interface compatibility
+    end
+
+    u = u0
+    convfail = false
+    for i in 2:length(ts)
+        uprev = u
+        t = ts[i]
+        nlf = isinplace(f) ? (out, u, p) -> f(out, u, uprev, p, t) : (u, p) -> f(u, uprev, p, t)
```

Comment:
is the compiler always smart enough to optimize this closure? It probably is these days. I would've just made a callable type, but I think this got better. I guess we can just patch this if we find a case where it does allocate.